### PR TITLE
Module2

### PR DESCRIPTION
**Pull request by StanislavMilavitsky**

### CODE REVIEW

File: api/src/java/com/epam/esm/configuration/SpringConfig.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        return dataSource;
+    }
+
+    /**
+     * Set properties for database
+     *
+     * @return bean driver manager by database
+     */
+   /* @Bean
+    @Profile("prod")
```
Comment (volynetspavel) [inline api/src/java/com/epam/esm/configuration/SpringConfig.java]: why this is commented?

File: api/src/java/com/epam/esm/controller/GiftCertificateController.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+     * Delete gift certificate by id.
+     *
+     * @param id the id
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @DeleteMapping("/{id}")
+    public ResponseEntity<String> delete(@PathVariable(name = "id") Long id) throws ServiceException {
+        long result = giftCertificateService.delete(id);
+        return result != -1L ? ResponseEntity.ok("Delete successful!") : ResponseEntity.ok("Delete unsuccessful!");
```
Comment (volynetspavel) [inline api/src/java/com/epam/esm/controller/GiftCertificateController.java]: maybe better create boolean variable the long?

File: api/src/java/com/epam/esm/controller/GiftCertificateController.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    /**
+     * Find gift certificate by tag name.
+     *
+     * @param tagName the tag name
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @GetMapping("/byName/{name}")
```
Comment (volynetspavel) [inline api/src/java/com/epam/esm/controller/GiftCertificateController.java]: is name part of url or name is parameter in url?
check please how to create names for url https://hackernoon.com/restful-api-designing-guidelines-the-best-practices-60e1d954e7c9

File: api/src/java/com/epam/esm/controller/TagController.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    /**
+     * Add tag.
+     *
+     * @param tagName the tag name
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @PostMapping(value = "/", consumes = "application/json")
```
Comment (volynetspavel) [inline api/src/java/com/epam/esm/controller/TagController.java]: is it necessary to have value for one post method? 
why do you need consumes there? does it work without consumes defenition?

File: api/src/java/com/epam/esm/exception/ControllerException.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package com.epam.esm.exception;
+
+public class ControllerException extends Exception {
```
Comment (volynetspavel) [inline api/src/java/com/epam/esm/exception/ControllerException.java:6]: what is this exception doing?

File: core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    private final GiftCertificateConverter giftCertificateConverter;
+
+    public GiftCertificateServiceImpl(GiftCertificateDAO giftCertificateDAO, GiftCertificateConverter giftCertificateConverter) {
+        this.giftCertificateDAO = giftCertificateDAO;
+        this.giftCertificateConverter = giftCertificateConverter;
+    }
+
+    @Override
+    public GiftCertificateDTO find(Long id) throws ServiceException {
+        GiftCertificate giftCertificate = giftCertificateDAO.read(id);
```
Comment (volynetspavel) [inline core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: maybe better check on null before converting to dto?

File: core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            logger.error(exception);
+            throw new ServiceException(exception);
+        }
+
+    }
+
+    @Override
+    public List<GiftCertificateDTO> searchByNameOrDescription(String part) throws ServiceException {
+            List<GiftCertificate> giftCertificates = giftCertificateDAO.searchByNameOrDescription(part);
+        if (giftCertificates != null) {
```
Comment (volynetspavel) [inline core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: what if list is not null but empty?

File: core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        }
+    }
+
+    @Override
+    public List<GiftCertificateDTO> sortByName(SortType sortType) throws ServiceException {
+        List<GiftCertificate> giftCertificates = giftCertificateDAO.sortByName(sortType);
+        if (giftCertificates != null){
+            return giftCertificates.stream().map(giftCertificateConverter::toDTO).collect(Collectors.toList());
+        } else {
+            logger.error("Sort by name exception!");
```
Comment (volynetspavel) [inline core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: it's not quite an understandable message for the exception, in the message should be the reason: why we we have such exception.

File: core/src/test/java/com/epam/esm/dao/impl/TagDAOImplTest.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Test
+    void read() {
+        Tag actual = tagDAO.read(1L);
+        Tag expected = new Tag(1L, "sport");
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void delete() {
```
Comment (volynetspavel) [inline core/src/test/java/com/epam/esm/dao/impl/TagDAOImplTest.java]: there is should be also negative tests

File: core/src/main/java/com/epam/esm/dao/impl/GiftCertificateDAOImpl.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    @Override
+    public GiftCertificate read(Long id){
+            return jdbcTemplate.queryForObject(READ_GIFT_CERTIFICATE_BY_ID_SQL, new BeanPropertyRowMapper<>(GiftCertificate.class), id);
+    }
+
+    @Override
+    public long update(GiftCertificateDTO giftCertificateDTO)  {
+         int rows = jdbcTemplate.update(UPDATE_GIFT_CERTIFICATE_SQL, giftCertificateDTO.getName(), giftCertificateDTO.getDescription(),
+                giftCertificateDTO.getPrice(), giftCertificateDTO.getDuration(), giftCertificateDTO.getLastUpdateDate(), giftCertificateDTO.getId());
+        return rows > 0L ? giftCertificateDTO.getId() : -1L;
```
Comment (volynetspavel) [inline core/src/main/java/com/epam/esm/dao/impl/GiftCertificateDAOImpl.java]: there is should be returned updated entity, to see result, that your entity was updated

File: core/src/main/java/com/epam/esm/dao/impl/GiftCertificateDAOImpl.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+
+    @Override
+    public GiftCertificate read(Long id){
+            return jdbcTemplate.queryForObject(READ_GIFT_CERTIFICATE_BY_ID_SQL, new BeanPropertyRowMapper<>(GiftCertificate.class), id);
+    }
+
+    @Override
+    public long update(GiftCertificateDTO giftCertificateDTO)  {
```
Comment (volynetspavel) [inline core/src/main/java/com/epam/esm/dao/impl/GiftCertificateDAOImpl.java]: as I see your dao works only with entity, not with dto. would be better use there entity too

**Review by volynetspavel — CHANGES_REQUESTED** [review]

File: api/src/java/com/epam/esm/configuration/SpringConfig.java (hunk 1 above)
Comment (StanislavMilavitsky) [reply api/src/java/com/epam/esm/configuration/SpringConfig.java]: Deleted methods with annotation  Profile

File: api/src/java/com/epam/esm/controller/GiftCertificateController.java (hunk 2 above)
Comment (StanislavMilavitsky) [reply api/src/java/com/epam/esm/controller/GiftCertificateController.java]: In response, I preferred to return the id, after which it was possible to work with it if that need

File: api/src/java/com/epam/esm/controller/GiftCertificateController.java (hunk 3 above)
Comment (StanislavMilavitsky) [reply api/src/java/com/epam/esm/controller/GiftCertificateController.java]: Understand and fix

File: api/src/java/com/epam/esm/controller/TagController.java (hunk 4 above)
Comment (StanislavMilavitsky) [reply api/src/java/com/epam/esm/controller/TagController.java]: Value in method not necessary.
In this method i need consumes, I'm saying that the method should accept only json it is work without consumes but body of request may be all types

File: api/src/java/com/epam/esm/exception/ControllerException.java (hunk 5 above)
Comment (StanislavMilavitsky) [reply api/src/java/com/epam/esm/exception/ControllerException.java:6]: Fix java doc in this class. This exception response object of exception with message if body or parameters not valid

File: core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 7 above)
Comment (StanislavMilavitsky) [reply core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: FIxed and deleted

File: core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 8 above)
Comment (StanislavMilavitsky) [reply core/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: FIxed

File: core/src/test/java/com/epam/esm/dao/impl/TagDAOImplTest.java (hunk 9 above)
Comment (StanislavMilavitsky) [reply core/src/test/java/com/epam/esm/dao/impl/TagDAOImplTest.java]: Fix

File: core/src/main/java/com/epam/esm/dao/impl/GiftCertificateDAOImpl.java (hunk 10 above)
Comment (StanislavMilavitsky) [reply core/src/main/java/com/epam/esm/dao/impl/GiftCertificateDAOImpl.java]: It returns, next step in services methode update use method dao.read, after dao.update

File: core/src/main/java/com/epam/esm/dao/impl/GiftCertificateDAOImpl.java (hunk 11 above)
Comment (StanislavMilavitsky) [reply core/src/main/java/com/epam/esm/dao/impl/GiftCertificateDAOImpl.java]: This should be so that the method does not return null pointer exception

**Pull request merged**